test_conformance/subgroups/subhelpers.h

```diff
@@ -56,220 +199,957 @@ class subgroupsAPI {
+}
+
+// set scalar value to half scalar
+template <typename Ty>
+typename std::enable_if<TypeManager<Ty>::is_sb_scalar_type::value>::type
+set_value(Ty &lhs, const cl_ulong &rhs)
 {
-    static bool val(cl_device_id) { return true; }
-};
-template <> struct TypeCheck<cl_half>
+    lhs.data = rhs;
+}
+
+
+// compare for common vectors
+template <typename Ty>
+typename std::enable_if<TypeManager<Ty>::is_vector_type::value, bool>::type
+compare(const Ty &lhs, const Ty &rhs)
 {
-    static bool val(cl_device_id device)
+    const int size = sizeof(Ty) / sizeof(typename TypeManager<Ty>::scalar_type);
+    for (auto i = 0; i < size; ++i)
     {
-        return is_extension_available(device, "cl_khr_fp16");
+        // log_info("Values index i = %d got %lu expected %lu\n", i, lhs.s[i],
+        //         rhs.s[i]);
+        if (lhs.s[i] != rhs.s[i])
+        {
+            return false;
+        }
     }
-};
-template <> struct TypeCheck<double>
+    return true;
+}
+
+
+// compare for vectors 3
+template <typename Ty>
+typename std::enable_if<TypeManager<Ty>::is_sb_vector_size3::value, bool>::type
+compare(const Ty &lhs, const Ty &rhs)
 {
-    static bool val(cl_device_id device)
+    for (auto i = 0; i < 3; ++i)
     {
-        int error;
-        cl_device_fp_config c;
-        error = clGetDeviceInfo(device, CL_DEVICE_DOUBLE_FP_CONFIG, sizeof(c),
-                                (void *)&c, NULL);
-        if (error)
+        if (lhs.data.s[i] != rhs.data.s[i])
         {
-            log_info(
-                "clGetDeviceInfo failed with CL_DEVICE_DOUBLE_FP_CONFIG\n");
             return false;
         }
-        return c != 0;
     }
-};
+    return true;
+}
+
+// compare for half vectors
+template <typename Ty>
+typename std::enable_if<TypeManager<Ty>::is_sb_vector_type::value, bool>::type
+compare(const Ty &lhs, const Ty &rhs)
+{
+    const int size = sizeof(Ty) / sizeof(typename TypeManager<Ty>::scalar_type);
+    for (auto i = 0; i < size; ++i)
+    {
+        if (lhs.data.s[i] != rhs.data.s[i])
+        {
+            return false;
+        }
+    }
+    return true;
+}
+
+
+// compare for scalars
+template <typename Ty>
+typename std::enable_if<std::is_scalar<Ty>::value, bool>::type
+compare(const Ty &lhs, const Ty &rhs)
+{
+    return lhs == rhs;
+}
+
+// compare for scalar halfs
+template <typename Ty>
+typename std::enable_if<TypeManager<Ty>::is_sb_scalar_type::value, bool>::type
+compare(const Ty &lhs, const Ty &rhs)
+{
+    return lhs.data == rhs.data;
+}
+
+template <typename Ty> inline bool compare_ordered(const Ty &lhs, const Ty &rhs)
+{
+    return lhs == rhs;
+}
 
+template <>
+inline bool compare_ordered(const subgroups::cl_half &lhs,
+                            const subgroups::cl_half &rhs)
+{
+    return cl_half_to_float(lhs.data) == cl_half_to_float(rhs.data)
+        && !is_half_nan(lhs.data);
+}
+
+template <typename Ty>
+inline bool compare_ordered(const subgroups::cl_half &lhs, const int &rhs)
```

Comment:
It's a bit unusual to provide an int to float comparison, as opposed to making the caller reconcile the types. The equality comparison is done in float with an implicit lossy promotion of the int to float, so there will cases where this claims equality for differing integer and float values (ie when the integer has over 24 significant bits). If that's acceptable for the purposes of these tests, it should be noted; if not, some argument checking would need to be done to ensure that the int argument can be exactly converted to float.